Windows kube-proxy panics after a LoadBalancer service is created

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


#### What happened:
kube-proxy panics when a LoadBalancer service is created after a deployment it is associated with is created and the pods are fully running.

#### What you expected to happen:
kube-proxy does not panic.
#### How to reproduce it (as minimally and precisely as possible):
1) Create 2 Windows nodes
```
# oc get nodes -l kubernetes.io/os=windows -owide
NAME                                         STATUS   ROLES    AGE    VERSION                       INTERNAL-IP    EXTERNAL-IP   OS-IMAGE                         KERNEL-VERSION    CONTAINER-RUNTIME
ip-10-0-133-68.us-east-2.compute.internal    Ready    worker   123m   v1.20.0-1030+cac2421340a449   10.0.133.68    <none>        Windows Server 2019 Datacenter   10.0.17763.1817   docker://19.3.14
ip-10-0-140-118.us-east-2.compute.internal   Ready    worker   130m   v1.20.0-1030+cac2421340a449   10.0.140.118   <none>        Windows Server 2019 Datacenter   10.0.17763.1817   docker://19.3.14
```
2) Create a deployment, ensuring that the pods are scheduled on different nodes:
```
# cat WinWebServer.yaml 
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: win-webserver
  name: win-webserver
spec:
  selector:
    matchLabels:
      app: win-webserver
  replicas: 2
  template:
    metadata:
      labels:
        app: win-webserver
      name: win-webserver
    spec:
      tolerations:
      - key: "os"
        value: "Windows"
        Effect: "NoSchedule"
      containers:
      - name: windowswebserver
        image: mcr.microsoft.com/windows/servercore:ltsc2019
        imagePullPolicy: IfNotPresent
        command:
        - powershell.exe
        - -command
        - $listener = New-Object System.Net.HttpListener; $listener.Prefixes.Add('http://*:80/'); $listener.Start();Write-Host('Listening at http://*:80/'); while ($listener.IsListening) { $context = $listener.GetContext(); $response = $context.Response; $content='<html><body><H1>Windows Container Web Server</H1></body></html>'; $buffer = [System.Text.Encoding]::UTF8.GetBytes($content); $response.ContentLength64 = $buffer.Length; $response.OutputStream.Write($buffer, 0, $buffer.Length); $response.Close(); };
      nodeSelector:
        beta.kubernetes.io/os: windows
```
3) Wait for the pods to go into the `running` state. kube-proxy will not crash if the pods are not fully deployed
4) Create a LoadBalancer service for the deployment
```
NAME            TYPE           CLUSTER-IP       EXTERNAL-IP                                                              PORT(S)        AGE
win-webserver   LoadBalancer   172.30.200.235   a40403ddbf3da439289d9fc3b0875f5c-743076964.us-east-2.elb.amazonaws.com   80:30043/TCP   4h1m

# curl a40403ddbf3da439289d9fc3b0875f5c-743076964.us-east-2.elb.amazonaws.com
<html><body><H1>Windows Container Web Server</H1></body></html>

# cat WinLoadBalancer.yaml
apiVersion: v1
kind: Service
metadata:
  name: win-webserver
  labels:
    app: win-webserver
spec:
  ports:
  # the port that this service should serve on
  - port: 80
    targetPort: 80
  selector:
    app: win-webserver
  type: LoadBalancer
```
5) At this point kube-proxy should have crashed. If not, delete and recreate the LoadBalancer, this should trigger the panic.
#### Anything else we need to know?:
The kube-proxy exe was run as:
```
kube-proxy.exe --v=4 --proxy-mode=kernelspace --feature-gates=WinOverlay=true --hostname-override=ip-10-0-129-184.ec2.internal --kubeconfig=c:\k\kubeconfig --cluster-cidr=10.132.1.0/24 --log-dir=C:\var\log\kube-proxy\ --logtostderr=false --network-name=OVNKubernetesHybridOverlayNetwork --source-vip=10.132.1.170 --enable-dsr=false
```
Stdout/err shows: 
```
E0318 19:55:57.397147    3808 proxier.go:1137] Endpoint information not available for service default/win-webserver. Not applying any policy 
E0318 19:55:57.547539    3808 proxier.go:1188] Policy creation failed: hcnCreateLoadBalancer failed in Win32: The reference is invalid in the JSON document. (0x803b001c) {"Success":false,"Error":"The reference is invalid in the JSON document. A Reference is required","ErrorCode":2151350300} 
E0318 19:55:57.636353    3808 proxier.go:1188] Policy creation failed: hcnCreateLoadBalancer failed in Win32: The reference is invalid in the JSON document. (0x803b001c) {"Success":false,"Error":"The reference is invalid in the JSON document. A Reference is required","ErrorCode":2151350300} 
E0318 19:55:57.730110    3808 proxier.go:1188] Policy creation failKubernetes version: 1.20
Windows version: Windows Server 2019 10.0.17763.1817
ed: hcnCreateLoadBalancer failed in Win32: The reference is invalid in the JSON document. (0x803b001c) {"Success":false,"Error":"The reference is invalid in the JSON document. A Reference is required","ErrorCode":2151350300} 
E0318 19:55:57.883467    3808 proxier.go:1188] Policy creation failed: hcnCreateLoadBalancer failed in Win32: The reference is invalid in the JSON document. (0x803b001c) {"Success":false,"Error":"The reference is invalid in the JSON document. A Reference is required","ErrorCode":2151350300} 
E0318 19:55:57.972311    3808 proxier.go:1188] Policy creation failed: hcnCreateLoadBalancer failed in Win32: The reference is invalid in the JSON document. (0x803b001c) {"Success":false,"Error":"The reference is invalid in the JSON document. A Reference is required","ErrorCode":2151350300} 
E0318 19:55:58.472063    3808 proxier.go:1188] Policy creation failed: hcnCreateLoadBalancer failed in Win32: The reference is invalid in the JSON document. (0x803b001c) {"Success":false,"Error":"The reference is invalid in the JSON document. A Reference is required","ErrorCode":2151350300} 
time="2021-03-18T19:56:00Z" level=info msg="currentVersion.Major < versionRange.MinVersion.Major: 9, 12" 
time="2021-03-18T19:56:00Z" level=info msg="currentVersion.Major < versionRange.MinVersion.Major: 9, 10" 
time="2021-03-18T19:56:00Z" level=info msg="currentVersion.Major < versionRange.MinVersion.Major: 9, 12"
time="2021-03-18T19:56:00Z" level=info msg="currentVersion.Major < versionRange.MinVersion.Major: 9, 13"
time="2021-03-18T19:56:00Z" level=info msg="currentVersion.Major < versionRange.MinVersion.Major: 9, 10"
time="2021-03-18T19:56:00Z" level=info msg="currentVersion.Major < versionRange.MinVersion.Major: 9, 12"
time="2021-03-18T19:56:00Z" level=info msg="currentVersion.Major < versionRange.MinVersion.Major: 9, 10"
time="2021-03-18T19:56:00Z" level=info msg="currentVersion.Major < versionRange.MinVersion.Major: 9, 12"
time="2021-03-18T19:56:00Z" level=info msg="currentVersion.Major < versionRange.MinVersion.Major: 9, 13"
time="2021-03-18T19:56:00Z" level=info msg="currentVersion.Major < versionRange.MinVersion.Major: 9, 10"
panic: runtime error: invalid memory address or nil pointer dereference 
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x205c0dc]

goroutine 100 [running]:
k8s.io/kubernetes/pkg/proxy/winkernel.(*Proxier).syncProxyRules(0xc0000cd500)
        /remote-source/build/windows-machine-config-operator/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/proxy/winkernel/proxier.go:1124 +0x9dc
k8s.io/kubernetes/pkg/util/async.(*BoundedFrequencyRunner).tryRun(0xc00028c5a0)
        /remote-source/build/windows-machine-config-operator/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/async/bounded_frequency_runner.go:292 +0xbe
k8s.io/kubernetes/pkg/util/async.(*BoundedFrequencyRunner).Loop(0xc00028c5a0, 0xc0000820c0)
        /remote-source/build/windows-machine-config-operator/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/async/bounded_frequency_runner.go:203 +0x21b
k8s.io/kubernetes/pkg/proxy/winkernel.(*Proxier).SyncLoop(0xc0000cd500)
        /remote-source/build/windows-machine-config-operator/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/proxy/winkernel/proxier.go:752 +0x72
created by k8s.io/kubernetes/cmd/kube-proxy/app.(*ProxyServer).Run
        /remote-source/build/windows-machine-config-operator/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kube-proxy/app/server.go:778 +0x855
 ```
Full kube-proxy logs:
[kube-proxy.exe.ERROR.txt](https://github.com/kubernetes/kubernetes/files/6167134/kube-proxy.exe.ERROR.txt)
[kube-proxy.exe.INFO.txt](https://github.com/kubernetes/kubernetes/files/6167135/kube-proxy.exe.INFO.txt)
[kube-proxy.exe.WARNING.txt](https://github.com/kubernetes/kubernetes/files/6167136/kube-proxy.exe.WARNING.txt)



#### Environment:
- Kubernetes version (use `kubectl version`): 1.20
- Cloud provider or hardware configuration: AWS
- OS (e.g: `cat /etc/os-release`): Windows server 2019
- Kernel (e.g. `uname -a`): 10.0.17763.1817
- Install tools: OpenShift 4.7.3
- Network plugin and version (if this is a network-related bug): Hybrid OVN
- Others:

--- a/pkg/proxy/winkernel/proxier.go
+++ b/pkg/proxy/winkernel/proxier.go
@@ -354,7 +354,7 @@ func newSourceVIP(hns HostNetworkService, network string, ip string, mac string,
 
 func (ep *endpointsInfo) Cleanup() {
 	Log(ep, "Endpoint Cleanup", 3)
-	if ep.refCount != nil {
+	if !ep.GetIsLocal() && ep.refCount != nil {
 		*ep.refCount--
 
 		// Remove the remote hns endpoint, if no service is referring it
@@ -1157,10 +1157,10 @@ func (proxier *Proxier) syncProxyRules() {
 			} else {
 				// We only share the refCounts for remote endpoints
 				ep.refCount = proxier.endPointsRefCount.getRefCount(newHnsEndpoint.hnsID)
+				*ep.refCount++
 			}
 
 			ep.hnsID = newHnsEndpoint.hnsID
-			*ep.refCount++
 
 			Log(ep, "Endpoint resource found", 3)
 		}

